File: src/main/java/com/milaboratory/mixcr/util/ClonesAlignmentRanges.java
package com.milaboratory.mixcr.util;

import com.milaboratory.core.Range;
import com.milaboratory.core.alignment.Alignment;
import com.milaboratory.core.mutations.Mutation;
import com.milaboratory.core.sequence.NucleotideSequence;
import com.milaboratory.mixcr.basictypes.Clone;
import com.milaboratory.mixcr.basictypes.VDJCHit;
import com.milaboratory.mixcr.trees.CloneWrapper;
import io.repseq.core.GeneType;
import io.repseq.core.ReferencePoint;
import org.apache.commons.math3.util.Pair;

import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.function.Function;
import java.util.stream.Collectors;

import static io.repseq.core.ReferencePoint.CDR3Begin;
import static io.repseq.core.ReferencePoint.CDR3End;

public class ClonesAlignmentRanges {
    private final List<Range> commonRanges;
    private final GeneType geneType;

    public ClonesAlignmentRanges(List<Range> commonRanges, GeneType geneType) {
        this.commonRanges = commonRanges;
        this.geneType = geneType;
    }

    public List<Range> getCommonRanges() {
        return commonRanges;
    }

    public static Range CDR3Sequence1Range(VDJCHit hit, Alignment<NucleotideSequence> alignment) {
        int from = getRelativePosition(hit, CDR3Begin);
        int to = getRelativePosition(hit, CDR3End);
        if (from == -1 && to == -1) {
            return null;
        }
        if (from == -1) {
            from = alignment.getSequence1Range().getLower();
        }
        if (to == -1) {
            to = alignment.getSequence1Range().getUpper();
        }
        return new Range(from, to);
    }

    private static int getRelativePosition(VDJCHit hit, ReferencePoint referencePoint) {
        return hit.getGene().getPartitioning().getRelativePosition(hit.getAlignedFeature(), referencePoint);
    }

    public Range cutRange(Range range) {
        return commonRanges.stream()
                .filter(it -> it.intersectsWith(range))
                .findFirst()
                .orElseThrow(() -> new IllegalArgumentException("range is not represented in common ranges"));
    }

    public boolean containsMutation(int mutation) {
        int position = Mutation.getPosition(mutation);
        return commonRanges.stream().anyMatch(it -> it.contains(position));
    }

    public boolean containsCloneWrapper(CloneWrapper clone) {
        return Arrays.stream(clone.getHit(geneType).getAlignments())
                .map(Alignment::getSequence1Range)
                .allMatch(range -> commonRanges.stream().anyMatch(range::contains));
    }

    public boolean containsClone(Clone clone) {
        return Arrays.stream(clone.getBestHit(geneType).getAlignments())
                .map(Alignment::getSequence1Range)
                .allMatch(range -> commonRanges.stream().anyMatch(range::contains));
    }

    public static <T> ClonesAlignmentRanges commonAlignmentRanges(List<T> clones, double minPortionOfClones, GeneType geneType, Function<T, VDJCHit> hitSupplier) {
        if (minPortionOfClones < 0.5) {
            throw new IllegalArgumentException("if minPortionOfClones < 0.5 than there may be ranges with intersections that exist in all clones");
        }
        // allow excluding minimum one clone in small clusters, but not breaking algorithm condition to filtering on more than a half of clones
        int threshold = Math.max((int) Math.floor(clones.size() * minPortionOfClones), (int) Math.ceil(clones.size() / 2.0));
        Map<Range, Long> rangeCounts = clones.stream()
                .flatMap(clone -> {
                    VDJCHit bestHit = hitSupplier.apply(clone);
                    return Arrays.stream(bestHit.getAlignments())
                            .map(alignment -> {
                                Range CDR3Range = CDR3Sequence1Range(bestHit, alignment);
                                if (CDR3Range == null) {
                                    return alignment.getSequence1Range();
                                } else {
                                    List<Range> withoutCDR3 = alignment.getSequence1Range().without(CDR3Range);
                                    if (withoutCDR3.size() != 1) {
                                        throw new IllegalStateException();
                                    } else {
                                        return withoutCDR3.get(0);
                                    }
                                }
                            });
                })
                .collect(Collectors.groupingBy(Function.identity(), Collectors.counting()));
        List<Range> rangesWithMinPortionOfClones = rangeCounts.keySet().stream()
                .map(range -> Pair.create(
                        range,
                        rangeCounts.entrySet().stream()
                                .filter(e -> e.getKey().contains(range))
                                .mapToLong(Map.Entry::getValue)
                                .sum()
                ))
                .filter(it -> it.getSecond() >= threshold)
                .map(Pair::getKey)
                .collect(Collectors.toList());
        List<Range> result = rangesWithMinPortionOfClones.stream()
                .filter(range -> rangesWithMinPortionOfClones.stream()
                        .filter(it -> !it.equals(range))
                        .noneMatch(it -> it.contains(range)))
                .collect(Collectors.toList());
        boolean noIntersections = result.stream().noneMatch(range -> result.stream()
                .filter(it -> !it.equals(range))
                .anyMatch(range::intersectsWith)
        );
        if (!noIntersections) {
            throw new IllegalStateException("there are intersections");
        }
        return new ClonesAlignmentRanges(result, geneType);
    }
}

File: src/main/kotlin/com/milaboratory/mixcr/util/ClonesAlignmentRanges.kt
@file:Suppress("FunctionName", "LocalVariableName")

package com.milaboratory.mixcr.util

import com.milaboratory.core.Range
import com.milaboratory.core.alignment.Alignment
import com.milaboratory.core.mutations.Mutation
import com.milaboratory.core.sequence.NucleotideSequence
import com.milaboratory.mixcr.basictypes.Clone
import com.milaboratory.mixcr.basictypes.VDJCHit
import com.milaboratory.mixcr.trees.CloneWrapper
import io.repseq.core.GeneType
import io.repseq.core.ReferencePoint
import kotlin.math.ceil
import kotlin.math.floor
import kotlin.math.max

class ClonesAlignmentRanges(
    val commonRanges: Array<Range>,
    private val geneType: GeneType
) {
    fun cutRange(range: Range): Range = commonRanges.first { it.intersectsWith(range) }

    fun containsMutation(mutation: Int): Boolean {
        val position = Mutation.getPosition(mutation)
        return commonRanges.any { it.contains(position) }
    }

    fun containsCloneWrapper(clone: CloneWrapper): Boolean =
        clone.getHit(geneType).alignments
            .all { alignment -> commonRanges.any { other -> other in alignment.sequence1Range } }

    fun containsClone(clone: Clone): Boolean =
        clone.getBestHit(geneType).alignments
            .all { alignment -> commonRanges.any { other -> other in alignment.sequence1Range } }

    companion object {
        fun CDR3Sequence1Range(hit: VDJCHit, alignment: Alignment<NucleotideSequence>): Range? {
            var from = getRelativePosition(hit, ReferencePoint.CDR3Begin)
            var to = getRelativePosition(hit, ReferencePoint.CDR3End)
            if (from == -1 && to == -1) {
                return null
            }
            if (from == -1) {
                from = alignment.sequence1Range.lower
            }
            if (to == -1) {
                to = alignment.sequence1Range.upper
            }
            return Range(from, to)
        }

        private fun getRelativePosition(hit: VDJCHit, referencePoint: ReferencePoint): Int =
            hit.gene.partitioning.getRelativePosition(hit.alignedFeature, referencePoint)

        fun <T> commonAlignmentRanges(
            clones: List<T>,
            minPortionOfClones: Double,
            geneType: GeneType,
            hitSupplier: (T) -> VDJCHit
        ): ClonesAlignmentRanges {
            require(minPortionOfClones >= 0.5) { "if minPortionOfClones < 0.5 than there may be ranges with intersections that exist in all clones" }
            // allow excluding minimum one clone in small clusters, but not breaking algorithm condition to filtering on more than a half of clones
            val threshold = max(floor(clones.size * minPortionOfClones).toInt(), ceil(clones.size / 2.0).toInt())
            val rangeCounts = clones
                .flatMap { clone: T ->
                    val bestHit = hitSupplier(clone)
                    bestHit.alignments.map { alignment ->
                        when (val CDR3Range = CDR3Sequence1Range(bestHit, alignment)) {
                            null -> alignment.sequence1Range
                            else -> {
                                val withoutCDR3 = alignment.sequence1Range.without(CDR3Range)
                                when (withoutCDR3.size) {
                                    1 -> withoutCDR3[0]
                                    else -> throw IllegalStateException()
                                }
                            }
                        }
                    }
                }
                .groupingBy { it }.eachCount()
            val rangesWithMinPortionOfClones = rangeCounts
                .mapValues { (range) ->
                    range to rangeCounts
                        .filterKeys { key -> range in key }
                        .values
                        .sum()
                }
                .filterValues { it.second >= threshold }
                .keys
            val result = rangesWithMinPortionOfClones.filter { range ->
                (rangesWithMinPortionOfClones - range).none { other -> range in other }
            }
            val noIntersections = result.none { range ->
                (result - range).any { other -> range.intersectsWith(other) }
            }
            check(noIntersections) { "there are intersections" }
            return ClonesAlignmentRanges(result.toTypedArray(), geneType)
        }
    }
}
